### ISSUE TITLE
bug(error-logger): `console.error` may be undefined

### ISSUE THREAD

**Issue by AmitMY** [angular/angular#15833]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
On a Galaxy S4 device (running Android 5.0.1), in a web application using ionic, *sometimes*, it would throw the following error:
> ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'apply' of undefined(…)defaultErrorLogger

This error originates from: https://github.com/angular/angular/blob/4.0.x/packages/core/src/errors.ts#L36
(Source map points me there)

Which is very interesting, as why would the `console.error` not exist? who knows.

**Expected behavior**
Some kind of a `try/catch` in the `defaultErrorLogger` method, or some check if `console.log` exists.

**Minimal reproduction of the problem with instructions**
I don't know, it happens sometimes, on a device.

**What is the motivation / use case for changing the behavior?**
The error logger should not throw errors..

**Please tell us about your environment:**
```
    Angular version: 4.0.0
    Cordova CLI: 6.5.0
    Ionic CLI Version: 3.0.0-beta.5
    ios-deploy version: Not installed
    ios-sim version: Not installed
    OS: Windows 10
    Node Version: v6.9.1
    Xcode version: Not installed
```
Device for errors:
```
Device: Galaxy S4
OS: Android 5.0.1
```

* **Angular version:** 4.0.0

* **Browser:** Android 5.0.1 Web Browser
 
* **Language:** TypeScript 2.2.1

* **Node (for AoT issues):** `node --version` =   v6.9.1

**Comment by tbosch** [comment]
Thanks for the bug report. However, we need to have a reproduction for this, otherwise we can't investigate it...

**Comment by pkozlowski-opensource** [comment]
@tbosch I think that this might be IE9 only where console is undefined when dev tools are not opened.

**Comment by AmitMY** [comment]
@pkozlowski-opensource While that is a nice note, it's kind of unrelated, as for the fact that in this case, `console` is defined, while `console.error` is not.

But, if there is a chance of `console` or `console.error` to not be defined, I think that there should be a `try/catch` there.

@tbosch I can't manage to reproduce it myself, as it happened only once that I have noticed, but it definitely pointed to that piece of code.

**Comment by Toxicable** [comment]
@AmitMY are you able to make a small project that we can run where this could happen?

**Comment by AmitMY** [comment]
@Toxicable I am unable to reproduce it myself (as said above), because I am not sure what triggered an error and when exactly I need to trigger an error for the `ErrorHandler` to log an error before the `console.error` method being available.

Clarification: After I got the error, when manually doing `console.error` (remote debugging), it did exist.

**Comment by tbosch** [comment]
So what is the underlying error that should have been logged?

**Comment by AmitMY** [comment]
@tbosch I am not sure, as if ever it gets to there, it just says:
> ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'apply' of undefined(…)defaultErrorLogger

So I don't know what the error is supposed to be.

---

I understand I am not much help in this situation, just sharing the fact I had a bug, that I can't reproduce, on an old device, originating from Angular (really regarding the `console` object)

**Comment by AmitMY** [comment]
@tbosch There is this issue that has very similar strokes to it, and can reproduce constantly.
https://github.com/driftyco/ionic/issues/11269
@frankkoenigstein mind sharing more info? what is your device and Android version?

**Comment by frankkoenigstein** [comment]
I'm running prod version on my Nexus 5X with Anroid 7.1.2.
But if I run the code generated by ionic run android --prod with ionic serve I face the same error in my chrome dev browser on my pc. For me it seems not related to any device.

If I find the time I'll provide two sample repositories.

**Comment by mifkys** [comment]
@tbosch I have same issue on Android < 4.4 (tested on 4.1, 4.2, 4.3) with Ionic app:
> 12-04 06:10:48.296 8600-8600/io.ionic.starter D/SystemWebChromeClient: file:///android_asset/www/build/vendor.js: Line 1703 : ERROR
> 12-04 06:10:48.296 8600-8600/io.ionic.starter E/Web Console: ERROR at file:///android_asset/www/build/vendor.js:1703

> 1703:     console.error.apply(console, values);

Environment:

>     @ionic/cli-utils  : 1.19.0
>     ionic (Ionic CLI) : 3.19.0
> 
> global packages:
> 
>     cordova (Cordova CLI) : 7.0.1
> 
> local packages:
> 
>     @ionic/app-scripts : 3.1.2
>     Cordova Platforms  : android 6.2.3
>     Ionic Framework    : ionic-angular 3.9.2
> 
> System:
> 
>     Android SDK Tools : 26.1.1
>     Node              : v8.9.0
>     npm               : 5.5.1
>     OS                : Windows 10
> 
> Environment Variables:
> 
>     ANDROID_HOME : C:\dev\android-sdk

Repository: https://github.com/mifkys/angular-android-console-error (also added debug apk file with this problem). To reproduce run apk file on Android 4.1-4.3 or 
start new Ionic project
add android platform
`ionic cordova run android`

Need help!

**Comment by tytskyi** [comment]
Some environments do not have `console` api when not using developer tools.

It is possible to mock console methods for those.

1. Install https://www.npmjs.com/package/console-polyfill 
2. `import "console-polyfill";` in `main.ts` or `polyfill.ts`;

**Comment by mifkys** [comment]
@tytskyi maybe I didn't understand but install console-polyfill and import it in main.ts not solve problem. Error is the same

**Comment by tytskyi** [comment]
@mifkys then it does not help, sorry

**Comment by mifkys** [comment]
any news? How can I temporary resolve this issue?

**Comment by pkozlowski-opensource** [comment]
I'm going to close this issue given that:
* we don't have reproduce scenario
* we didn't see new reports / mentions of this error in ~3 years. 

If anyone still bumps into this / similar problem, please open a new issue with a reroduce scenario.
